### ISSUE TITLE
New Tooltips

### ISSUE THREAD

**Issue by mrcthms** [Shopify/polaris-react-archive#7989]
## Tooltip

The Tooltips are undergoing a revamp, with three distinct intentional use cases for the Tooltip. These are:

### Definition Tooltip

Definition of done:

- The popover moves with the mouse, with a little springiness btwn mouse position and tooltip (to be fine-tuned)
- The popover appears immediately on hover
- On click, the popover transitions to a fixed position (left aligned, 8px up from the hovered text block)
- The "click-and-stick" transition has a microanimation to make it feel like it snaps in place
- Tooltip will adjust positioning to avoid going out of the browser window viewable area
- The component has keyboard a11y
- The component has voice-over a11y

[Prototype](https://tooltip-prototype.shopifycloud.com/1)
Related ticket: https://github.com/Shopify/web/issues/78345

### Icon tooltip:

Definition of done:

- Tooltip appears on hover after 300ms
- The positioning is above the icon, but will appear below the icon if there is no room in the browser window
- The microanimation is a 4px translation in the y-direction as the opacity goes to 100%
- The tooltip component has a prop for showing keyboard shortcut if there is one
- If the mouse moves directly to another icon, that new icon's tooltip will be visible immediately
- The component has keyboard a11y
- The component has voice-over a11y

[Prototype](https://tooltip-prototype.shopifycloud.com/1/iconTooltip)
Related ticket: https://github.com/Shopify/web/issues/78346

### Action tooltip:

The Action tooltip is still being finalised, so isn't ready to be included in scope for this initial Tooltip work. You can follow along with the progress [here](https://github.com/Shopify/web/issues/78351).

---

**We are proposing updating the Tooltip component in an additive, non-breaking way to cater for the Definition tooltip and Icon tooltip (along with the Action tooltip when it ready for build), by adding a `mode` prop which will control the behaviour of the Tooltip.**

## Required components

One of the key UX patterns with the new Tooltip is that if a merchant is navigating from one Tooltip activator to another immediately, then the Tooltip related to the second activator should show immediately, not after 300ms as currently happens. In order to do this, we propose including a new EphemeralPresenceProvider, which will be a manager for determining which ephemeral components are currently visible on screen. This will allow all Tooltips access to this shared context to control the transition duration of the appearance animation. We will start by adding functionality for the Tooltip, but we see this as also being extended for components such as Toasts and BulkActions, so that components can adapt if multiple, disparate components are present simultaneously.

This EphemeralPresenceProvider will also be utilised by the upcoming [Abbreviation tooltips](https://github.com/Shopify/web/issues/78347), although we propose that this kind of tooltip should be a separate component due to it's unique use case that doesn't have much crossover with the regular Tooltip.

### Definition tooltip

These tooltips exist to offer additional contextual information to a merchant without requiring them to leave the page. They should be used for content that doesn’t need to be present at all times.

https://tooltip-prototype.shopifycloud.com/1/definition

Example for default callout card:

### Icon tooltip

These tooltips exist for instances where the merchant is most likely familiar with the content. 
This tooltip doesn’t provide new information. It just expands the information that already exists.

https://tooltip-prototype.shopifycloud.com/1/iconTooltip

## Props

List props and type declarations:

- **mode** (optional): `"icon" | "definition"` – controls the behaviour of the Tooltip. If no `mode` prop is given, the Tooltip behaves as currently exists today.

### Definition tooltip:

These tooltips exist to offer additional contextual information to a merchant without requiring them to leave the page. They should be used for content that doesn’t need to be present at all times.

### Icon tooltip:

These tooltips exist for instances where the merchant is most likely familiar with the content. This tooltip doesn’t provide new information. It just expands the information that already exists.

### Definition tooltip:

A concise definition that offers the context users need to understand what specifically is being measured. Keep them as short as possible, up to 2 sentences.   Use only when they provide enough value that a merchant won’t need to go to help docs. 

### Icon tooltip:

When augmenting an icon, the content of the tooltip must simply state the action that the icon represents. If icons appear together in a set (e.g. in the rich text editor) then every icon should have a tooltip. If the icon has a keyboard shortcut available, you may use the tooltip to reinforce this using a subdued colour. 

## EphemeralPresenceProvider

The EphemeralPresenceProvider will be a context provider that wraps the entire application, exposing an object to consumers to determine whether an instance of a component is currently present on screen. This is needed for the reasons explained above, where if a merchant navigates from Tooltip activator to Tooltip activator, the display should be instant.

Initially, it will return an object with a single key, comprising of a boolean value, but over time we see this Provider supporting multiple ephemeral components, such as the new Truncation tooltip, Bulk Actions, and Toasts, amongst others.

**Comment by johanstromqvist** [comment]
Thanks @mrcthms! The lists of intentions are great – much appreciated!

Interaction design looks good! The one thing that stuck out to me was the when mouse clicking the definition trigger and I get a focus state on the trigger plus toggling the positioning of the tooltip. Is that behaviour intentional? If so, what's the rationale?

<img width="543" alt="image" src="https://user-images.githubusercontent.com/38329124/210860413-174bd646-a95c-4067-bd35-9838a777ced2.png">

**Comment by mrcthms** [comment]
@dGoligorsky Can you chime in on this? Thanks!

**Comment by dGoligorsky** [comment]
@johanstromqvist @mrcthms 

If I understand the question correctly— why does the tooltip have a fixed position when clicking on the trigger (e.g. "Total sales")?

When the tooltip is open and in a fixed position, a merchant can mouse into the tooltip to highlight and copy text (e.g. if merchant wasn't familiar with the term "gross sales" and wanted to search it.)

**Comment by mrcthms** [comment]
@dGoligorsky I think the question was around the focus state being present when clicking on the Tooltip to toggle the sticking positioning. 

Currently, for Popovers, which this functionality mimics more than today's Tooltips, the focus is immediately given to the first focusable item within the [Popover](https://polaris.shopify.com/components/popover) content, so we don't see the focus state around the activator. For our definition tooltips, there will be no focusable element within the Tooltip content, however to ensure screen readers can navigate to the content within the Tooltip to read it, I feel like we should automatically focus the merchant to the container of the Tooltip. Do you have any aversion to that?

**Comment by dGoligorsky** [comment]
@mrcthms — No aversion to that!

To be clear, the merchant can't click on the popover, since it will be following the mouse. 
They'll click on the text ("Total sales") and you're recommending the focus state is on the popover.
Which sounds great to me!

**Comment by johanstromqvist** [comment]
Thanks! I was mostly curious about the purpose of the behaviour. Your thoughts on both position toggle and focus state sounds good to me. You got this!

**Comment by tatianau** [comment]
In catalogs, we have a use case when a tooltip looks like a `definition tooltip` visually and works somewhat as an `action tooltip` functionally. Or rather, in our case, it’s technically a Popover component as we needed the ability to add custom markup (a link) and the ability to focus/click on it when it’s open, so we went with the Popover as the Tooltip component was either only supporting strings as content at that point or some other reasons I can't remember (cc @oksanashopify ).


https://user-images.githubusercontent.com/15054990/214927540-f31a9156-ed28-4eb9-96c8-d9de5b1bbeb8.mp4



Will the new Tooltip have support for cases like ours?

---
Update: IIRC, @oksanashopify and I weren't able to make the focus/click on the content to work, even with dismissOnMouseOut={false}, and we had trouble making the focus return to the activator and it was going to the next interactive element instead - looks like those were the reasons we chose Popover over Tooltip where we managed to do all of it.

**Comment by LauraAubin** [comment]
@tatianau the Tooltip content was recently updated to support a React Node element so you can now pass in your link! 🙌

You should also be able to focus/click on the content when it's open with it's existing functionality, if that doesn't work you might need to set the `dismissOnMouseOut` prop to `false`.

**Comment by LauraAubin** [comment]
Regarding the underline underneath the tooltip activator, would this become part of the default tooltip, or built into the `definition` variation of the `mode` prop?

**Comment by oksanashopify** [comment]
Hi @LauraAubin! Just want to second the information @tatianau provided, is this the right way to create a Tooltip with Link inside: [sendbox](https://codesandbox.io/s/naughty-borg-gt72y2?file=/App.js)?  For me focus does not go to the Polaris Link element even when I  set dismissOnMouseOut prop to false. I am missing something? 
Thank you!

**Comment by mrcthms** [comment]
@tatianau @LauraAubin @oksanashopify Tooltips should never contain any actionable element, such as a button or link, as per the [guidelines](https://polaris.shopify.com/components/tooltip). If you've got any interactive element, it has to be within a Popover rather than a Tooltip

**Comment by mrcthms** [comment]
@alex-page Commenting here with thoughts about how to proceed as finding a meeting time for the both of us will be v. tricky with our current timezones. Also adding in the comments from the Slack thread for posterity:

> Alex Page
>  [14 days ago](https://shopify.slack.com/archives/C4Y8N30KD/p1675160462195849?thread_ts=1675158869.154159&cid=C4Y8N30KD)
> Thanks Marc.
> I am unsure about the approach. We likely want to decouple some parts here to make Tooltip component composable.
> Adding multiple modes will add complexity to the Tooltip and encourage addition of extra modes in the future. This usually means more css, properties and if statements changing what renders. Is there another way we could build this functionality without having to create multiple modes in Tooltip?
>
> My gut feeling is that a larger breaking change would be better for the system. Then slotting in modes.

Here's the additions needed for both the Icon and Definition tooltips:

| Tooltip type | Feature description |
| --- | --- |
| Icon | The microanimation is a 4px translation in the y-direction as the opacity goes to 100% |
| Icon | The tooltip component has a prop for showing keyboard shortcut if there is one |
| Icon | If the mouse moves directly to another icon, that new icon's tooltip will be visible immediately |
| Definition | The tooltip moves with the mouse, with a little springiness btwn mouse position and tooltip (to be fine-tuned) |
| Definition | On click, the tooltip transitions to a fixed position (left aligned, 8px up from the hovered text block) |
| Definition | The "click-and-stick" transition has a microanimation to make it feel like it snaps in place |

Looking at the direction we're going in for the Action tooltip, that will be an extension of the Popover due to it containing interactive content, so that wouldn't be a concern of the Tooltip. So these are the only updates we are making to this component as part of the Tooltip work – although it would be safe to assume the direction we opt for here will influence how we update the Popover for the Action tooltip/popover.

#### So here's what I'm thinking could be done to accommodate this:

I can't really think of a good argument for gating the additions to the `Icon` tooltip behind a `mode` prop, rather than them just being improvements to the original Tooltip component. Anecdotally, the vast majority of our tooltips used in the Admin would fall into the "icon" category, but implementing this within `web` using a `mode="icon"` prop would mean a massive manual step of adding `mode="icon"` to each instance of the Tooltip. By improving the base Tooltip with these updates, the related PR within `web` would be much smaller, as each Tooltip would have this behaviour automatically.

For the Definition tooltip, everything net-new for that flavour of Tooltip is related to the click-and-stick behaviour, so we could add a solitary `hasClickAndStick` prop to the Tooltip which adds this functionality in. The `hasClickAndStick` prop would enable the mouse-tracking functionality, the fixed positioning on click, and the updating of the microanimation on appear/hide/stick.

We add new props to our components all the time, and this to me feels like a valid approach. We do away with the new `mode` prop for fear of obfuscating the declarative behaviour of the Tooltip, and instead only add a `hasClickAndStick` prop to the Tooltip for those instances where we want the Tooltip to behave like a Definition tooltip.

By approaching it in this way, the need for a wholesale rewrite of the component is nullified, and the Quality crew can still remain the stewards of this work, and we can get it out to our merchants in a relatively small timeframe.

## TLDR

- We remove the `mode` prop entirely
- All updates that would fall into the "icon" mode are made to the base Tooltip as generic improvements to the component
- New `hasClickAndStick` prop handled behaviour required by the Definition tooltip

cc @Shopify/polaris-eng for thoughts and questions, also.

**Comment by dGoligorsky** [comment]
Thanks for this @mrcthms!! 
You and Alex know best on implementation, but one little comment is that the prop name might be more logical as `hasMouseFollow` or something more related to that functionality than `hasClickAndStick`? Just a thought 🤷

**Comment by alex-page** [comment]
@mrcthms thank you for moving the conversation here.

I am aligned with this new approach. I feel much better about building the core functionality into the tooltip and adding the `hasClickAndStick` property. I think this is the best long term solution for tooltips.
